packages/lib/contracts/application/App.sol

```diff
@@ -1,32 +1,92 @@
 
 /**
- * @title BaseApp
- * @dev Abstract base contract for upgradeable applications.
+ * @title App
+ * @dev Contract for upgradeable applications.
  * It handles the creation and upgrading of proxies.
  */
-contract BaseApp is Ownable {
+contract App is Ownable {
   /**
    * @dev Emitted when a new proxy is created.
    * @param proxy Address of the created proxy.
    */
   event ProxyCreated(address indexed proxy);
 
   /**
-   * @dev Constructor function
+   * @dev Emitted when a package dependency is changed in the application.
+   * @param providerName Name of the package that changed.
+   * @param package Address of the package associated to the name.
+   * @param version Version of the package in use.
    */
-  constructor() public {}
+  event PackageChanged(string providerName, address package, string version);
 
   /**
-   * @dev Abstract function to return the implementation provider for a given package name.
+   * @dev Tracks a package in a particular version, used for retrieving implementations
+   */
+  struct ProviderInfo {
+    Package package;
+    string version;
+  }
+  
+  /**
+   * @dev Maps from dependency name to a tuple of package and version
+   */
+  mapping(string => ProviderInfo) internal providers;
+
+  /**
+   * @dev Constructor function.
+   */
+  constructor() public { }
+
+  /**
+   * @dev Returns the provider for a given package name, or zero if not set.
    * @param packageName Name of the package to be retrieved.
-   * @return The implementation provider for the package.
+   * @return The provider.
    */
-  function getProvider(string packageName) public view returns (ImplementationProvider);
+  function getProvider(string packageName) public view returns (ImplementationProvider) {
+    ProviderInfo storage info = providers[packageName];
+    if (address(info.package) == address(0)) return ImplementationProvider(0);
+    return info.package.getVersion(info.version);
+  }
+
+  /**
+   * @dev Returns information on a package given its name.
+   * @param packageName Name of the package to be queried.
+   * @return A tuple with the package address and pinned version given a package name, or zero if not set
+   */
+  function getPackage(string packageName) public view returns (Package, string) {
+    ProviderInfo storage info = providers[packageName];
+    return (info.package, info.version);
+  } 
+
+  /**
+   * @dev Sets a package in a specific version as a dependency for this application. 
+   * Requires the version to be present in the package.
+   * @param packageName Name of the package to set or overwrite.
+   * @param package Address of the package to register.
+   * @param version Version of the package to use in this application.
+   */
+  function setPackage(string packageName, Package package, string version) public onlyOwner {
+    require(package.hasVersion(version), "The requested version must be registered in the given package");
+    providers[packageName] = ProviderInfo(package, version);
```

Comment:
should we check that there was no other package already set under that name?

Reply:
I'm thinking if we should force the user to unset the previous package in that case

Reply:
I wouldn't, as this allows the user to use this method to also change the version of a package if they want to do so. And I'd reduce the number of txs to interact with the App as much as possible.